MantidPlot/src/Spectrogram.cpp

```diff
@@ -237,6 +239,25 @@ MantidQt::API::QwtRasterDataMD *Spectrogram::dataFromWorkspace(
   auto dim1 = workspace->getDimension(1);
   Mantid::coord_t minX(dim0->getMinimum()), maxX(dim0->getMaximum()),
       minY(dim1->getMinimum()), maxY(dim1->getMaximum());
+
+  // A MatrixWorkspace can be ragged. Make sure the x axis covers all
+  // histograms.
+  auto matrixWorkspace =
+      boost::dynamic_pointer_cast<const Mantid::API::MatrixWorkspace>(
+          workspace);
+  if (matrixWorkspace) {
+    for (size_t iHisto = 0; iHisto < matrixWorkspace->getNumberHistograms();
+         ++iHisto) {
+      auto &x = matrixWorkspace->x(iHisto);
+      if (x.front() < minX) {
+        minX = static_cast<Mantid::coord_t>(x.front());
+      }
+      if (x.back() > maxX) {
+        maxX = static_cast<Mantid::coord_t>(x.back());
+      }
+    }
+  }
```

Comment:
This is good. But can we move it to a separate private method? This one already does quite a lot.